doc/src/manual/interfaces.md

```diff
@@ -371,3 +371,190 @@ If you are defining an array type that allows non-traditional indexing (indices
+computing result types and container sizes, and just do everything manually. For example,
+you can convert a `UnitRange{Int}` `rng` to a `UnitRange{BigInt}` with `big.(rng)`; the definition
+of this method is approximately
+
+```julia
+Broadcast.broadcast(::typeof(big), rng::UnitRange) = big(first(rng)):big(last(rng))
+```
+
+This exploits Julia's ability to dispatch on a particular function type. (This kind of
+explicit definition can indeed be necessary if the output container does not support `setindex!`.)
+You can optionally choose to implement the actual broadcasting yourself, but allow
+the internal machinery to compute the container type, element type, and indices by specializing
+
+```julia
+Broadcast.broadcast(::typeof(somefunction), r::Broadcast.Result{ContainerType,ElType,<:Tuple}, As...)
+```
+
+### [Writing binary broadcasting rules](@id writing-binary-broadcasting-rules)
+
+Binary rules look something like this:
+
+    Broadcast.rule(::Type{Primary}, ::Type{Secondary}) = Primary
+
+This would indicate that `Primary` has precedence over `Secondary`.
+Generally you should only define one argument order, because internal machinery will test
+both orders.
+The result does not have to be one of the input arguments, it could be a third type.
+For example, you could imagine defining
+
+    Broadcast.rule(::Type{Ref}, ::Type{Tuple}) = Vector
+
+so that a `Ref` and a `Tuple` broadcast to a `Vector`. (`Ref` is handled a bit
+differently than this by internal machinery, so this is just for the purposes
+of illustration.)
+
+While there are exceptions, in general you may find it simpler if the arguments avoid
+subtyping, e.g., in binary `rule`s use `Type{Primary}` rather than `Type{<:Primary}`.
+The main motivation for this advice is that subtyping can lead to ambiguities or conflicts
+for types that are subtypes of other types.
+As a consequence, if you define binary `rule` methods, when defining corresponding unary
+`rule` methods if possible you should discard type parameters, i.e.,
+
+    Broadcast.rule{<:Primary} = Primary
+
+rather than
+
+    Broadcast.rule{P} where P<:Primary = P
+
+`Broadcast` defines several internal types that can assist in writing binary rules:
+- `Broadcast.Scalar` for objects that act like scalars
+- `Broadcast.BottomArray{N}` for array types that haven't declared specialized broadcast implementations
+
+`BottomArray` stores the dimensionality as a type parameter (thus violating the advice above)
+to support specialized array types that have fixed dimensionality
+requirements. `BottomArray` "loses" to other array types that have specialized broadcasting
+rules because of the following method:
+
+    Broadcast.rule(::Type{<:AbstractArray{T,N}}) where {T,N} = BottomArray{N}
+    Broadcast.rule(::Type{A}, ::Type{BottomArray{N}}) where {A<:AbstractArray,N} = A
+
+This applies whenever the container type returned by unary `rule` is a subtype of
+`AbstractArray`. (The type of the *value* is irrelevant.)
```

Comment:
You mean "element type is irrelevant"?

Reply:
Sorry, I mean the type of `X` is irrelevant as long as `rule(typeof(X))` returns a "signaling type" that *is* a subtype of `AbstractArray` that is *not* `BottomArray`.

Reply:
Got it. Expanding a bit more would be helpful.